source/common/router/scoped_rds.cc

```diff
@@ -283,6 +283,11 @@ bool ScopedRdsConfigSubscription::addOrUpdateScopes(
       throw EnvoyException("route_configuration_name is empty.");
     }
     const std::string scope_name = scoped_route_config.name();
+    if (const auto& scope_info_iter = scoped_route_map_.find(scope_name);
+        scope_info_iter != scoped_route_map_.end() &&
+        scope_info_iter->second->configHash() == MessageUtil::hash(scoped_route_config)) {
+      continue;
```

Comment:
Can we compare the config directly here? It is possible that two different config has the same name and hash.
(scope_info_iter->second->configProto() == scoped_route_config) should work.

Reply:
https://github.com/envoyproxy/envoy/blob/a9c3a9663539097d7e1ec67036e2306cb553bdc5/source/common/protobuf/utility.h#L228

Reply:
Relevant issue: #6252
FWIW, I think the main reason `Equivalent` is not typically used is because the config proto isn't being kept.

Reply:
hmm, we definitely saves the source config for cluster and RDS route tables. 

For RDS it's in ConfigInfo: 
```
  struct ConfigInfo {
    // A reference to the currently loaded route configuration. Do not hold this reference beyond
    // the caller of configInfo()'s scope.
    const Protobuf::Message& config_;

    // The discovery version that supplied this route. This will be set to "" in the case of
    // static clusters.
    const std::string version_;
  };

```

for CDS it's in ClusterData: 

```
  struct ClusterData : public ClusterManagerCluster {
    ClusterData(const envoy::config::cluster::v3::Cluster& cluster_config,
                const uint64_t cluster_config_hash, const std::string& version_info,
                bool added_via_api, ClusterSharedPtr&& cluster, TimeSource& time_source)
        : cluster_config_(cluster_config), config_hash_(cluster_config_hash),
          version_info_(version_info), added_via_api_(added_via_api), cluster_(std::move(cluster)),
          last_updated_(time_source.systemTime()) {}

    bool blockUpdate(uint64_t hash) { return !added_via_api_ || config_hash_ == hash; }

    const envoy::config::cluster::v3::Cluster cluster_config_;
```